scripts/lib/installer-win32-gen.tsx

```diff
@@ -0,0 +1,238 @@
+
+/** @jsx Element.new */
+
+import crypto from 'crypto';
+import fs from 'fs';
+import path from 'path';
+
+/**
+ * Element is a class for interpreting JSX; we only need the bare basics to
+ * generate a valid XML as input to the WiX toolchain.
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars -- Can't detect @jsx usage.
+export class Element {
+  constructor(tag: string, attribs: Record<string, string>, ...children: (Element | string)[]) {
+    this.tag = tag;
+    this.attribs = attribs;
+    this.children = children;
+  }
+
+  /**
+   * Create a new element; this is used by the TypeScript JSX support.
+   */
+  static new(tag: string, attribs: Record<string, string> | null, ...children: (Element | Element[] | string)[]) {
+    return new Element(tag, attribs ?? {}, ...children.flat());
+  }
+
+  tag: string;
+  attribs: Record<string, string>;
+  // eslint-disable-next-line no-use-before-define -- ESLint gets confused by recursive references
+  children: (Element | string)[];
+
+  /** Convert the Element to serialized XML. */
+  toXML(indent = 0) {
+    const indentString = (new Array(indent + 1)).join(' ');
+    let result = `${ indentString }<${ this.tag }`;
+
+    for (const [key, value] of Object.entries(this.attribs)) {
+      result += ` ${ key }="${ value }"`;
+    }
+    if (this.children.length < 1) {
+      result += '/>\n';
+    } else {
+      result += '>';
+      if (this.children.some(c => c instanceof Element)) {
+        result += '\n';
+      }
+      for (const child of this.children) {
+        if (typeof child === 'string') {
+          // For text content of elements, always use CDATA.
+          result += `<![CDATA[${ child }]]>`;
+        } else if (child instanceof Element) {
+          result += child.toXML(indent + 2);
+        } else {
+          throw new TypeError(`Don't know how to serialize ${ child } (type ${ typeof child })`);
+        }
+      }
+      if (this.children.some(c => c instanceof Element)) {
+        result += indentString;
+      }
+      result += `</${ this.tag }>${ '\n' }`;
+    }
+
+    return result;
+  }
+}
+
+// When rendering, the JSX tag name is supposed to be the name of a component
+// that's passed to the first argument of React.createElement; so we need plain
+// constants for every element we use.
+const Component = 'Component';
+const ComponentGroup = 'ComponentGroup';
+const ComponentGroupRef = 'ComponentGroupRef';
+const Directory = 'Directory';
+const File = 'File';
+const Fragment = 'Fragment';
+const Shortcut = 'Shortcut';
+const ShortcutProperty = 'ShortcutProperty';
+
+/**
+ * A structure representing the files (but not subdirectories) within a
```

Comment:
I find the comment a little confusing, it says `but not subdirectories` however, I see a directories array (line: 89) within the directory type, isn't that meant to represent sub-dirs?

Reply:
Right, the comment was out of date; updated (to include subdirectories).